### DISCUSSION TITLE
Plugin is not available

### DISCUSSION THREAD

**Discussion by slavaschmidt** [scalameta/nvim-metals#497] [Q&A]
Hello! I'm probably doing something wrong but I can't figure out what exactly.

Metals is not available in nvim after installation.

Prerequisites are met, and the packer says the plugin is installed, but no metals commands are available in nvim.

When I try to update metals via Telescope ( :Telescope metals commands > Update Metals), I get the following error message:

```
...ite/pack/packer/start/nvim-metals/Lua/metals/install.lua:125: attempt to index local 'config' (a nil value)
```

Metals is installed:

```
 cs setup --yes --apps ammonite,cs,coursier,scala,scalac,sbt,sbtn,scalafmt,metals
Checking if a JVM is installed
Found a JVM installed under /Users/slasch/Library/Java/JavaVirtualMachines/corretto-17.0.5/Contents/Home.

Checking if ~/Library/Application Support/Coursier/bin is in PATH

Checking if the standard Scala applications are installed
  Found ammonite
  Found cs
  Found coursier
  Found scala
  Found scalac
  Installed sbt
  Installed sbtn
  Installed scalafmt
  Found metals
```

and nvim-metals also installed (last line):

```
  ✓ sharkdp/fd: already up to date
 ✓ mfussenegger/nvim-dap: already up to date
 ⟳ nvim-telescope/telescope.nvim: pulling updates
 ✓ hrsh7th/cmp-nvim-lsp: already up to date
 ✓ andymass/vim-matchup: already up to date
 ✓ hrsh7th/cmp-buffer: already up to date
 ✓ hrsh7th/cmp-cmdline: already up to date
 ✓ tpope/vim-dispatch: already up to date
 ✓ lewis6991/gitsigns.nvim: already up to date
 ✓ hrsh7th/nvim-cmp: already up to date
 ✓ folke/tokyonight.nvim: already up to date
 ✓ glacambre/firenvim: already up to date
 ✓ w0rp/ale: already up to date
 ✓ aserowy/tmux.nvim: already up to date
 ✓ hrsh7th/vim-vsnip: already up to date
 ✓ hrsh7th/vim-vsnip-integ: already up to date
 ✓ haorenW1025/completion-nvim: already up to date
 ✓ BurntSushi/ripgrep: already up to date
 ✓ nvim-lua/plenary.nvim: already up to date
 ✓ tjdevries/colorbuddy.vim: already up to date
 ✓ nvim-treesitter/nvim-treesitter: already up to date
 ✓ wbthomason/packer.nvim: already up to date
 ✓ hrsh7th/cmp-path: already up to date
 ✓ scalameta/nvim-metals: already up to date
```

What else could I try? Thank you!

**Comment by ckipp01** [comment]
Hey! What's your setup look like? Basically are you creating a metals config and passing it into `initialize_or_attach`? What's the autocmd that you have set up to launch nvim-metals look like?

**Reply by slavaschmidt** [reply]
Thank you very much for the quick response!

The `plugins.lua` executes `require('init_metals')` and the very end.

The full contents of `init_metals.lua` where `initialize_or_attach` is called at the end are :
```
local api = vim.api
local cmd = vim.cmd
local map = vim.keymap.set

local metals_config = require("metals").bare_config()

metals_config.settings = {
  showImplicitArguments = true,
  excludedPackages = { "akka.actor.typed.javadsl", "com.github.swagger.akka.javadsl" },
}

metals_config.init_options.statusBarProvider = "on"

local dap = require("dap")

dap.configurations.scala = {
  {
    type = "scala",
    request = "launch",
    name = "RunOrTest",
    metals = {
      runType = "runOrTestFile",
      --args = { "firstArg", "secondArg", "thirdArg" }, -- here just as an example
    },
  },
  {
    type = "scala",
    request = "launch",
    name = "Test Target",
    metals = {
      runType = "testTarget",
    },
  },
}

metals_config.on_attach = function(client, bufnr)
  require("metals").setup_dap()
end

local nvim_metals_group = api.nvim_create_augroup("nvim-metals", { clear = true })
api.nvim_create_autocmd("FileType", {
  pattern = { "scala", "sbt", "java" },
  callback = function()
    require("metals").initialize_or_attach(metals_config)
  end,
  group = nvim_metals_group,
})
```

**Reply by ckipp01** [reply]
And just to make sure, with this configuration if you:

1. Open a Scala file
2. Do a `:MetalsInstall`

You get the error you mentioned above?

**Reply by slavaschmidt** [reply]
Great, thanks a lot for the hint! One step closer to the working configuration :)

When I opened a single scala file, the Metals commands became available! 

Now, with a single scala file open, I can issue Metals Commands, for example, Run Doctor (please see below).

The problem still is opening an SBT project. For example, if I start `nvim` in the project folder without parameters or with `build.sbt` as a parameter, the Metals commands are still available, but they do nothing. Even `Run Doctor` does nothing at all. 



This is the output of the doctor for a single file:
```
┌─────────────────────── Metals Doctor ───────────────────────┐
│## Metals Info                                               │
│  - Metals Server version: 0.11.9                            │
│  - Build server currently being used is Bloop v1.5.4.       │
│  - Metals Java: 17.0.5 from Amazon.com Inc. located at /    │
│Users/slasch/Library/Java/JavaVirtualMachines/corretto-17.0. │
│5/Contents/Home                                              │
│                                                             │
│These are the installed build targets for this workspace.    │
│One build target corresponds to one class path. For example,  │
│normally one sbt project maps to two build targets: main and │
│test.                                                        │
│                                                             │
│## Build Targets                                             │
│                                                             │
│### aimasch                                                  │
│  - target type: Scala 3.2.0                                 │
│  - compilation: ✅                                          │
│  - diagnostics: ✅                                          │
│  - interactive: ✅                                          │
│  - semanticdb: ✅                                           │
│  - debugging: ✅                                            │
│  - java: ✅                                                 │
│                                                             │
└─────────────────────────────────────────────────────────────┘

```

**Reply by ckipp01** [reply]
> if I start nvim in the project folder without parameters 

So when you start nvim without any parameter keep in mind you're just opening an empty buffer basically, so you aren't opening a filetype that is triggering Metals to attach. So this part is expected.

Opening a `build.sbt` file however _should_ cause the autocmd to trigger and attatch Metals. On a fresh sbt project if you directly open a `build.sbt` you _should_ be prompted to import the build. Is this happeing?

**Reply by slavaschmidt** [reply]
> So when you start nvim without any parameter keep in mind you're just opening an empty buffer basically, so you aren't opening a filetype that is triggering Metals to attach. So this part is expected.

This totally makes sense, thanks for the clarification. Old Coc habits tricked my understanding.

Opening `build.sbt` does nothing though. The `:MetalsInfo` works and shows an information popup, but unfortunately neither `:MetalsRunDoctor` nor `:MetalsImportBuild` do anything. 

Thank you!

**Reply by ckipp01** [reply]
> Opening build.sbt does nothing though.

This is odd. At this point everything should pretty much work. Can you check in your workspace if a `.metals/metals.log` was created? And if so, is there anything funky in there?

Also is this project just a simple hello world project?

**Reply by slavaschmidt** [reply]
> This is odd. At this point everything should pretty much work. Can you check in your workspace if a .metals/metals.log was created? And if so, is there anything funky in there?

Yep, the log is there and there is nothing suspicious in there. 

Interestingly enough, the `:MetalsRunDoctor` started to work for `build.sbt` after I truncated the log file 😮 

At this point, I believe this is just me doing things wrong or related to my environment. I'll experiment a bit looking into the log file. 

Thank you very much for your help!

**Reply by slavaschmidt** [reply]
Sorry for bothering you again with this. 
Yet another question: is it expected that "*.worksheet.sc" files can be run without having a build definition? 
Specifically, my simple test `object Hello extends App { println("hello world") }` compiled just fine, but I'm getting the `LSP[metals] There is nothing to run or test in the current file.` message if I try to run it

**Reply by ckipp01** [reply]
Hey @slavaschmidt worksheets don't need to be "ran". Upon save a worksheet is automatically evaluated and the results should be returned and shown to you like so:
<img width="472" alt="Screenshot 2022-11-21 at 11 52 14" src="https://user-images.githubusercontent.com/13974112/203032440-060df06c-9fa4-42c8-9af7-03cc003073a5.png">

**Reply by slavaschmidt** [reply]
that did the trick, huge thanks one more time!
